pattisdr/osf.io
# [Migration Required][QA Prep Required] Feature/Registration Responses Migration [ENG-842]

⭐️ Majority of this PR is a test to keep behavior from regressing.  Deeply nested registration test metadata is pretty lengthy, which comprises the bulk of the lines of this PR.

## Purpose 

Now that registration schemas have been flattened into schema blocks, the responses to these schema questions need to be likewise flattened.

## Changes

- Adds `registration_responses` fields and `registration_responses_migrated` (temporary) fields to `DraftRegistration` and `Registration` models
- Adds a migration to pull the relevant information from `DraftRegistration.registration_metadata` and `Registration.registered_meta` into `DraftRegistration.registration_responses` and `Registration.registration_responses` in local development - hoisting all nested answers to the top level. Given a `registration_response_key`, we want to be able to do `registration_responses[registration_response_key]` to get at the answer.
- On staging/prod, a management command will need to be run to incrementally populate these columns. 
- Adds `registration_responses` to the DraftRegistration and Registration serializers 
   - On the DraftRegistration serializer,  if you write to `registration_responses`, it is expanded and written to `registration_metadata`.  The reverse is also true, if you write to `registration_metadata`, it is flattened and written to `registration_responses`. 

## Example - open-ended registration
```json
"registered_meta": {
    "summary": {
           "comments": [],
           "value": "hello",
           "extra": []
      }
  },
 "registration_responses": {
    "summary": "hello"
}
```

files are stored like this:
```python
"recommended-hypothesis.question4a": [
                    {
                        "file_name": "Alphabet.txt",
                        "file_id": "5d6d25024d476c088fb8e03b"
                    },
                    {
                        "file_name": "Screen Shot 2019-08-30 at 9.04.01 AM.png",
                        "file_id": "5d6d25014d476c088fb8e038"
                    }
                ]
```

## QA Notes
- Migration prep - make several draft registrations and several registrations with different schemas.  Verify registration responses in API post-migration have deeply nested answers moved to highest level and all are present. 
- Start a draft registration or update a draft.  Verify that the API has `registration_responses` populated for that draft, which is a flattened version of the `registration_metadata`.  
- Create a registration. Verify that the API has `registration_responses` populated for that draft, which is a flattened version of the `registration_metadata`.  
## Documentation



## Side Effects



## Ticket

https://openscience.atlassian.net/browse/ENG-842
